src/video/videosurface.cpp

```diff
@@ -68,9 +70,14 @@ void VideoSurface::setSource(VideoSource *src)
     if (source == src)
         return;
 
-    unsubscribe();
-    source = src;
-    subscribe();
+    // To don't block GUI thread.
+    QtConcurrent::run([this, src]()
+    {
+        QMutexLocker l(&bigLock);

```

Comment:
>Do you have any proof?

First, here's what the docs state:
```text
The QtConcurrent namespace provides high-level APIs that make it possible to write multi-threaded programs without using low-level threading primitives.
…
```

That means, it is designed to not use any mutexes or semaphores! So if you feel the need to lock a mutex here, something's wrong… :)

Second, when using a mutex, you need to know exactly when to lock and unlock it to avoid very hard to fix races (deadlock)! In this specific case I wonder why it even works, because if you follow the flow into `CameraSource::subscribe` (and unsubscribe respectively), you'll find, that it should actually relock `bigLock`, causing a deadlock:

```c++
bool CameraSource::subscribe()
{
    QMutexLocker l{&biglock};

    if (!_isOpen)
    {
        ++subscriptions;
        return true;
    }

    if (openDevice())
    {
        ++subscriptions;
        return true;
    }
    else
    {
        while (device && !device->close()) {}
        device = nullptr;
        cctx = cctxOrig = nullptr;
        videoStreamIndex = -1;
        // Memfence so the stream thread sees a nullptr device
        std::atomic_thread_fence(std::memory_order_release);
        return false;
    }

}
```

This already runs in a non-blocking context! So the question here is: What blocks the main thread?

Anyway, I found another QtConcurrent::run in another place in code. Maybe you can take some inspriation from there:

```c++
ToxFriendCall::~ToxFriendCall()
{
    if (timeoutTimer)
        delete timeoutTimer;

    if (videoEnabled)
    {
        // This destructor could be running in a toxav callback while holding toxav locks.
        // If the CameraSource thread calls toxav *_send_frame, we might deadlock the toxav and CameraSource locks,
        // so we unsuscribe asynchronously, it's fine if the webcam takes a couple milliseconds more to poweroff.
        QtConcurrent::run([](){CameraSource::getInstance().unsubscribe();});
        …
    }
}
```

Probably it helps to move the `QtConcurrent::run` inside `Videsurface::subscribe(source)` and call it like so:

```c++
VideoSurface::setSource(src) {
    if (src != source) {
         unsubscribe();

         QtConcurrent::run([src]() {
             subscribe(src);
         }
         // note: you can also think of running subscribe asynchronously in general and move the `QtConcurrent::run` inside.
    }
}

Hope that helps.